File: backend/python_server/spiral_fit.py
# python_server/spiral_fit.py

import numpy as np

# --- アルゴリズム用の設定項目 ---
NUM_GENERATIONS = 80
NUM_CANDIDATES = 250
N_ELITES = 15
MUTATION_RATE = 0.25
MIN_A_THRESHOLD = 15.0
GOLDEN_B = 0.30635 
# ★★★ B_PENALTY_WEIGHT の固定値定義を削除 ★★★

def calculate_composition_score(candidate_params, points, image_shape):
    # ... (この関数は変更なし) ...
    h, w = image_shape
    cx, cy, a, b = candidate_params['cx'], candidate_params['cy'], candidate_params['a'], candidate_params['b']
    theta = np.linspace(-np.pi * 4, np.pi * 4, 200)
    r = a * np.exp(b * theta)
    x_fit, y_fit = cx + r * np.cos(theta), cy + r * np.sin(theta)
    valid_mask = (x_fit >= 0) & (x_fit < w) & (y_fit >= 0) & (y_fit < h)
    spiral_points = np.vstack((x_fit[valid_mask], y_fit[valid_mask])).T
    if len(spiral_points) < 10: return float('inf')
    total_distance = sum(np.min(np.sqrt(np.sum((spiral_points - p)**2, axis=1))) for p in points)
    return total_distance / len(points)

# ★★★ 関数の引数に `b_penalty_weight` を追加 ★★★
def optimize_spiral_with_golden_ratio(points, image_shape, b_penalty_weight):
    h, w = image_shape
    search_ranges = {'cx': [-w, 2*w], 'cy': [-h, 2*h], 'a': [10.0, 400.0], 'b': [0.1, 0.5]}
    
    candidates = []
    for _ in range(NUM_CANDIDATES):
        candidates.append({'cx': np.random.uniform(*search_ranges['cx']),'cy': np.random.uniform(*search_ranges['cy']),'a': np.random.uniform(*search_ranges['a']),'b': np.random.uniform(*search_ranges['b'])})

    best_overall_params, best_overall_score = {}, float('inf')

    for generation in range(NUM_GENERATIONS):
        scores = []
        for params in candidates:
            if params['a'] < MIN_A_THRESHOLD:
                scores.append(float('inf'))
                continue
            
            distance_score = calculate_composition_score(params, points, (h, w))
            b_penalty = (params['b'] - GOLDEN_B)**2
            
            # ★★★ 固定値の代わりに引数を使う ★★★
            final_score = distance_score + b_penalty_weight * b_penalty
            scores.append(final_score)
        
        # ... (以降の最適化ループは変更なし) ...
        sorted_candidates = sorted(zip(scores, candidates), key=lambda x: x[0])
        if sorted_candidates[0][0] < best_overall_score:
            best_overall_score, best_overall_params = sorted_candidates[0][0], sorted_candidates[0][1]
        elites = [c for s, c in sorted_candidates[:N_ELITES]]
        next_generation = elites[:]
        num_mutations = int(NUM_CANDIDATES * MUTATION_RATE)
        for _ in range(num_mutations): next_generation.append({'cx': np.random.uniform(*search_ranges['cx']), 'cy': np.random.uniform(*search_ranges['cy']), 'a': np.random.uniform(*search_ranges['a']), 'b': np.random.uniform(*search_ranges['b'])})
        num_offspring = NUM_CANDIDATES - len(next_generation)
        for i in range(num_offspring): parent = elites[i % len(elites)]; next_generation.append({'cx': np.random.normal(parent['cx'], w * 0.1), 'cy': np.random.normal(parent['cy'], h * 0.1), 'a': np.random.normal(parent['a'], 20.0), 'b': np.random.normal(parent['b'], 0.05)})
        candidates = next_generation
        if (generation + 1) % 10 == 0: print(f"世代 {generation+1}/{NUM_GENERATIONS}: ベストスコア = {best_overall_score:.4f} (b={best_overall_params.get('b', 0):.4f})")
    
    return best_overall_params

File: backend/rust_axum_server/src/spiral_fit.rs
use rand::Rng;
use rand::seq::SliceRandom;
use std::f64::consts::PI;
use std::cmp::Ordering;
use rayon::prelude::*;

const NUM_GENERATIONS: usize = 80;
const NUM_CANDIDATES: usize = 300;
const N_ELITES: usize = 15;
const MUTATION_RATE: f64 = 0.25;
const MIN_A_THRESHOLD: f64 = 15.0;
const GOLDEN_B: f64 = 0.30635;

#[derive(Clone, Debug)]
pub struct SpiralParams {
    pub cx: f64,
    pub cy: f64,
    pub a: f64,
    pub b: f64,
}

pub fn calculate_composition_score(params: &SpiralParams, points: &[[f64; 2]], image_shape: (u32, u32)) -> f64 {
    let (w, h) = (image_shape.0 as f64, image_shape.1 as f64);
    let theta_steps = 200;
    let max_theta = PI * 4.0;
    let min_theta = -PI * 4.0;
    
    let mut spiral_points = Vec::with_capacity(theta_steps);
    
    for i in 0..theta_steps {
        let t = min_theta + (max_theta - min_theta) * (i as f64 / (theta_steps - 1) as f64);
        let r = params.a * (params.b * t).exp();
        let x = params.cx + r * t.cos();
        let y = params.cy + r * t.sin();
        
        if x >= 0.0 && x < w && y >= 0.0 && y < h {
            spiral_points.push([x, y]);
        }
    }

    if spiral_points.len() < 10 {
        return f64::INFINITY;
    }

    let mut total_distance = 0.0;
    for point in points {
        let mut min_dist = f64::INFINITY;
        for sp in &spiral_points {
            let dist_sq = (sp[0] - point[0]).powi(2) + (sp[1] - point[1]).powi(2);
            if dist_sq < min_dist {
                min_dist = dist_sq;
            }
        }
        total_distance += min_dist.sqrt();
    }

    total_distance / points.len() as f64
}

pub fn optimize_spiral_with_golden_ratio(points: &[[f64; 2]], image_shape: (u32, u32), b_penalty_weight: f64) -> SpiralParams {
    let (w, h) = (image_shape.0 as f64, image_shape.1 as f64);
    let mut rng = rand::thread_rng();

    let mut candidates: Vec<SpiralParams> = (0..NUM_CANDIDATES).map(|_| {
        SpiralParams {
            cx: rng.gen_range(-w..2.0*w),
            cy: rng.gen_range(-h..2.0*h),
            a: rng.gen_range(10.0..400.0),
            b: rng.gen_range(0.1..0.5),
        }
    }).collect();

    let mut best_overall_params = candidates[0].clone();
    let mut best_overall_score = f64::INFINITY;

    for _generation in 0..NUM_GENERATIONS {
        let mut scored_candidates: Vec<(f64, SpiralParams)> = candidates.par_iter().map(|params| {
            if params.a < MIN_A_THRESHOLD {
                return (f64::INFINITY, params.clone());
            }

            let distance_score = calculate_composition_score(params, points, image_shape);
            let b_penalty = (params.b - GOLDEN_B).powi(2);
            let final_score = distance_score + b_penalty_weight * b_penalty;
            
            (final_score, params.clone())
        }).collect();

        scored_candidates.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(Ordering::Equal));

        if scored_candidates[0].0 < best_overall_score {
            best_overall_score = scored_candidates[0].0;
            best_overall_params = scored_candidates[0].1.clone();
        }

        let elites: Vec<SpiralParams> = scored_candidates.iter().take(N_ELITES).map(|x| x.1.clone()).collect();
        let mut next_generation = elites.clone();

        let num_mutations = (NUM_CANDIDATES as f64 * MUTATION_RATE) as usize;
        for _ in 0..num_mutations {
            next_generation.push(SpiralParams {
                cx: rng.gen_range(-w..2.0*w),
                cy: rng.gen_range(-h..2.0*h),
                a: rng.gen_range(10.0..400.0),
                b: rng.gen_range(0.1..0.5),
            });
        }

        let num_offspring = NUM_CANDIDATES - next_generation.len();
        for i in 0..num_offspring {
            let parent = &elites[i % elites.len()];
            next_generation.push(SpiralParams {
                cx: rng.gen::<f64>() * w * 0.1 + parent.cx,
                cy: rng.gen::<f64>() * h * 0.1 + parent.cy,
                a: rng.gen::<f64>() * 20.0 + parent.a,
                b: rng.gen::<f64>() * 0.05 + parent.b,
            });
        }
        candidates = next_generation;
    }

    best_overall_params
}
